content/academy/puppeteer_playwright/collecting_data.md

```diff
@@ -0,0 +1,311 @@
+
+> Most data collection cases involve looping through a list of items of some sort.
+
+There are two main ways to collect data with Playwright and Puppeteer:
+
+1. Directly in `page.evaluate()` and other evaluate functions such as `page.$$eval()`.
+2. In the Node.js context using a parsing library such as [Cheerio](https://www.npmjs.com/package/cheerio)
+
+Here is the base set up for our code, upon which we'll be building off of in this lesson:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const autoScroll = async (page) => {
+    await page.evaluate(async () => {
+        await new Promise((resolve) => {
+            let totalHeight = 0;
+            let distance = 100;
+            let interval = setInterval(() => {
+                const scrollHeight = document.body.scrollHeight;
+
+                window.scrollBy(0, distance);
+                totalHeight += distance;
+
+                if (totalHeight >= scrollHeight - window.innerHeight) {
+                    clearInterval(interval);
+                    resolve();
+                }
+            }, 200);
+        });
+    });
+};
+
+const sleep = (secs) => new Promise((r) => setTimeout(r, secs * 1000));
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForLoadState('networkidle');
+
+await autoScroll(page);
+
+// code will go here
+
+await sleep(10);
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteeer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const autoScroll = async (page) => {
+    await page.evaluate(async () => {
+        await new Promise((resolve) => {
+            let totalHeight = 0;
+            let distance = 100;
+            let interval = setInterval(() => {
+                const scrollHeight = document.body.scrollHeight;
+
+                window.scrollBy(0, distance);
+                totalHeight += distance;
+
+                if (totalHeight >= scrollHeight - window.innerHeight) {
+                    clearInterval(interval);
+                    resolve();
+                }
+            }, 200);
+        });
+    });
+};
+
+const sleep = (secs) => new Promise((r) => setTimeout(r, secs * 1000));
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForNetworkIdle();
+
+await autoScroll(page);
+
+// code will go here
+
+await sleep(10);
+
+await browser.close();
+</marked-tab>
+```
+
+Notice that we've written and are using a custom `autoScroll()` function, which will scroll to the very bottom of the page. We are doing this because the images of the products are lazy-loaded (they are only loaded once they've been scrolled into view).
```

Comment:
Instead of slowly scrolling to the bottom, maybe we could insta-scroll there? It would greatly simplify the function. Or the images wouldn't load this way?